src/rulesets/Base/mapreduce.jl

```diff
@@ -322,3 +325,122 @@ end
     end
     return dx
 end
+
+#####
+##### `foldl`
+#####
+
+# `foldl` guarantees to execute `f` in order, left to right. So it makes sense even when
+# this `f` is stateful, in which case the gradient must be calculated in the reverse order. 
+# The implementation aims to be efficient for both tuples and arrays. 
+# Note that it does not return a gradient for `init`.
+
+function rrule(
+        config::RuleConfig{>:HasReverseMode}, ::typeof(foldl), f::F, x::Union{AbstractArray, Tuple};
+        init=_InitialValue()
+    ) where {F}
+    list, start = if init !== _InitialValue()
+        x, init
+    else
+        _drop1(x), first(x)
+    end
+    hobbits = accumulate(list; init=(start, pi)) do (a,_), b
```

Comment:
I do not know how i feel about this
```suggestion
    # The Hobbit, or There and Back Again is a children's fantasy novel by English author J. R. R. Tolkien.
    # This is the primal result and the pullbacks (i.e. there and back again)
    hobbits = accumulate(list; init=(start, pi)) do (a,_), b
```

Reply:
Thinking about this more, I think even as a local variable this is to cute

`ȳ_pbs` is probably better